lib/postgres_v3_experimental.dart

```diff
@@ -0,0 +1,204 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+
+abstract class PgPool implements PgSession, PgSessionExecutor {
+  static Future<PgPool> open(
+    List<PgEndpoint> endpoints, {
+    PgSessionSettings? sessionSettings,
+    PgPoolSettings? poolSettings,
+  }) =>
+      throw UnimplementedError();
+
+  Future<R> withConnection<R>(
+    Future<R> Function(PgConnection connection) fn, {
+    PgSessionSettings? sessionSettings,
+  });
+}
+
+abstract class PgSession {
+  // uses extended query protocol
+  Future<PgStatement> prepare(
+    String sql, {
+    Object? /* String */ substitution,
+    Object? /* List<PgDataType> | Map<String, PgDataType> */ types,
+    Duration? timeout,
+  });
+
+  Future<PgResult> execute(
+    String sql, {
+    Object? /* String */ substitution,
+    Object? /* List<PgDataType> | Map<String, PgDataType> */ types,
+    Object? /* List<Object?|PgTypedParameter> | Map<String, Object?|PgTypedParameter> */ parameters,
+    Duration? timeout,
+  }) async {
+    if (substitution == null && types == null && parameters == null) {}
+    final stmt = await prepare(
+      sql,
+      substitution: substitution,
+      types: types,
+      timeout: timeout,
+    );
+    try {
+      return await stmt.run(parameters);
+    } finally {
+      await stmt.dispose();
+    }
+  }
+
+  Future<void> close();
+}
+
+abstract class PgSessionExecutor {
+  // TODO: also add retry options similarly to postgres_pool
+  Future<R> run<R>(Future<R> Function(PgSession session) fn);
+  Future<R> runTx<R>(Future<R> Function(PgSession session) fn);
+}
+
+abstract class PgConnection implements PgSession, PgSessionExecutor {
+  static Future<PgConnection> open(
+    PgEndpoint endpoint, {
+    PgSessionSettings? sessionSettings,
+  }) async =>
+      throw UnimplementedError();
+
+  PgChannels get channels;
+  PgMessages get messages;
+}
+
+abstract class PgResultStream implements Stream<PgResultRow> {
+  Future<int> get affectedRows;
+  Future<PgResultSchema> get schema;
+}
+
+abstract class PgStatement {
+  PgResultStream start(
```

Comment:
I like the idea of being able to run queries without caching all results in memory. But we also can't implement pauses/cancellations efficiently in this protocol, so I'm not sure if `Stream` is the right interface to use? Maybe this should be a callback, even if that generally feels less Dart-like? I'm not sure about this though. Perhaps it is enough to document that pauses are emulated for technical reasons.

Reply:
Not efficient, but first version could be just ignoring the rest of the results after cancelling the stream, but later on we could do something like this:
https://stackoverflow.com/questions/35319597/how-to-stop-kill-a-query-in-postgresql

Callbacks could work too, and potentially simpler to implement.

Reply:
If there's an eventual goal to support that I think streams are a good API. I was just concerned whether we could pull this off at all.

Reply:
I'm tempted to backpedal on the `Stream` design, and rather use a plain callback-based method, partly because it seems to me that the above-mentioned query killer is really a backend/connection killer.

Reply:
We can send an execute message with a fixed maximum row count (say 25). If there are more rows, postgres will send a `PortalSuspended` message informing us that more rows are available once we send another execute message.

So I think we could implement cancellations by just closing the portal and pauses by holding off sending execute messages.

----

Of course, we can also just add the non-streaming API first and then add a stream method once we have figured out how to implement it.

Reply:
Apparently I'm unaware of many details of the postgres protocols, it seems to be a great suggestion! For `Stream` we could use a `StreamController` internally and its `onPause`, `onResume` methods to control the flow. I think it wouldn't be much different from the callbacks.